File: macos/Sources/ClaudeUsageBar/DeviceSyncPayload.swift
import CryptoKit
import Foundation

struct DeviceSyncGeneral: Codable, Equatable {
    let pollingMinutes: Int
}

struct DeviceSyncAppearance: Codable, Equatable {
    let preferredProvider: String
    let menuBarStyle: String
    let primaryMetric: String
    let secondaryMetric: String
    let detailStyle: String
    let textSize: String
}

struct DeviceSyncNotifications: Codable, Equatable {
    let claudeSession: Int
    let claudeSevenDay: Int
    let claudeFable: Int
    let openAIWeekly: Int
    let openAIResetCredits: Int
    let cursorAPI: Int
    let cursorAuto: Int
    let cursorCredit: Int
}

struct DeviceSyncConnections: Codable, Equatable {
    let openAISessionToken: String?
    let cursorSessionToken: String?
    let elevenLabsAPIKey: String?

    var count: Int {
        [openAISessionToken, cursorSessionToken, elevenLabsAPIKey]
            .compactMap { $0 }
            .count
    }
}

struct DeviceSyncPayload: Codable, Equatable {
    static let currentVersion = 1
    static let validityDuration: TimeInterval = 10 * 60

    let version: Int
    let issuedAtEpochSeconds: Int64
    let expiresAtEpochSeconds: Int64
    let general: DeviceSyncGeneral?
    let appearance: DeviceSyncAppearance?
    let notifications: DeviceSyncNotifications?
    let connections: DeviceSyncConnections?

    init(
        issuedAt: Date = Date(),
        general: DeviceSyncGeneral? = nil,
        appearance: DeviceSyncAppearance? = nil,
        notifications: DeviceSyncNotifications? = nil,
        connections: DeviceSyncConnections? = nil
    ) {
        version = Self.currentVersion
        issuedAtEpochSeconds = Int64(issuedAt.timeIntervalSince1970)
        expiresAtEpochSeconds = Int64(
            issuedAt.addingTimeInterval(Self.validityDuration).timeIntervalSince1970
        )
        self.general = general
        self.appearance = appearance
        self.notifications = notifications
        self.connections = connections
    }
}

struct DevicePairingCode: Equatable {
    static let currentVersion = 2

    let sessionID: String
    let host: String
    let port: UInt16
    let desktopID: String
    let desktopName: String
    let desktopPublicKey: Data

    func encodedURLString() throws -> String {
        var components = URLComponents()
        components.scheme = "agentusagebar"
        components.host = "pair"
        components.path = "/v2"
        components.queryItems = [
            URLQueryItem(name: "v", value: String(Self.currentVersion)),
            URLQueryItem(name: "session", value: sessionID),
            URLQueryItem(name: "host", value: host),
            URLQueryItem(name: "port", value: String(port)),
            URLQueryItem(name: "desktop", value: desktopID),
            URLQueryItem(name: "name", value: desktopName),
            URLQueryItem(name: "key", value: desktopPublicKey.base64URLEncodedString())
        ]
        guard let value = components.string else {
            throw DeviceSyncError.invalidCode
        }
        return value
    }
}

struct DevicePairRequest: Codable {
    let sessionID: String
    let deviceID: String
    let deviceName: String
    let publicKey: String
}

struct DevicePairStartResponse: Codable {
    let status: String
    let confirmationCode: String
}

struct DevicePairPollResponse: Codable {
    let status: String
    let desktopID: String?
    let desktopName: String?
    let envelope: DeviceEncryptedEnvelope?
    let message: String?
}

struct DeviceEncryptedEnvelope: Codable, Equatable {
    let nonce: String
    let ciphertext: String
    let tag: String
}

struct DeviceStatusCommand: Codable {
    let action: String
    let issuedAtEpochSeconds: Int64
}

struct DeviceWipeAcknowledgement: Codable {
    let desktopID: String
    let deviceID: String
    let timestamp: Int64
    let proof: String
}

enum DeviceSyncCrypto {
    static let pairingInfo = Data("agentusagebar-device-pair-v2".utf8)
    static let statusInfo = Data("agentusagebar-device-status-v2".utf8)

    static func sharedSecret(
        desktopPrivateKey: P256.KeyAgreement.PrivateKey,
        devicePublicKey: Data
    ) throws -> SharedSecret {
        let publicKey = try P256.KeyAgreement.PublicKey(x963Representation: devicePublicKey)
        return try desktopPrivateKey.sharedSecretFromKeyAgreement(with: publicKey)
    }

    static func key(
        sharedSecret: SharedSecret,
        salt: String,
        info: Data
    ) -> SymmetricKey {
        sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: Data(salt.utf8),
            sharedInfo: info,
            outputByteCount: 32
        )
    }

    static func confirmationCode(sharedSecret: SharedSecret, sessionID: String) -> String {
        let key = key(sharedSecret: sharedSecret, salt: sessionID, info: pairingInfo)
        let bytes = key.withUnsafeBytes { Data($0) }
        let number = bytes.prefix(4).reduce(UInt32(0)) {
            ($0 << 8) | UInt32($1)
        } % 1_000_000
        return String(format: "%06u", number)
    }

    static func authenticationProof(
        sharedSecret: SharedSecret,
        salt: String,
        info: Data,
        message: String
    ) -> String {
        let symmetricKey = key(sharedSecret: sharedSecret, salt: salt, info: info)
        return Data(
            HMAC<SHA256>.authenticationCode(
                for: Data(message.utf8),
                using: symmetricKey
            )
        ).base64URLEncodedString()
    }

    static func seal<T: Encodable>(
        _ value: T,
        sharedSecret: SharedSecret,
        salt: String,
        info: Data
    ) throws -> DeviceEncryptedEnvelope {
        let data = try JSONEncoder.deviceSyncEncoder.encode(value)
        let symmetricKey = key(sharedSecret: sharedSecret, salt: salt, info: info)
        let box = try AES.GCM.seal(data, using: symmetricKey)
        return DeviceEncryptedEnvelope(
            nonce: Data(box.nonce).base64URLEncodedString(),
            ciphertext: box.ciphertext.base64URLEncodedString(),
            tag: box.tag.base64URLEncodedString()
        )
    }
}

enum DeviceSyncError: LocalizedError {
    case invalidCode
    case codeTooLarge
    case localNetworkUnavailable
    case serverUnavailable
    case pairingExpired
    case unknownDevice

    var errorDescription: String? {
        switch self {
        case .invalidCode:
            return "This is not a valid Agent Usage Bar device code."
        case .codeTooLarge:
            return "Could not create the pairing QR code."
        case .localNetworkUnavailable:
            return "Connect this Mac and phone to the same local network and try again."
        case .serverUnavailable:
            return "Could not start secure device pairing on this Mac."
        case .pairingExpired:
            return "This pairing request expired. Generate a new QR code."
        case .unknownDevice:
            return "This device is not registered with this Mac."
        }
    }
}

extension JSONEncoder {
    static let deviceSyncEncoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }()
}

extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    init?(base64URLEncoded value: String) {
        var base64 = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder != 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        self.init(base64Encoded: base64)
    }
}

File: android/app/src/main/java/com/agentusagebar/android/data/sync/DeviceSyncPayload.kt
package com.agentusagebar.android.data.sync

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.math.BigInteger
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.PublicKey
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.security.spec.PKCS8EncodedKeySpec
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

@Serializable
data class DeviceSyncGeneral(
    val pollingMinutes: Int,
)

@Serializable
data class DeviceSyncAppearance(
    val preferredProvider: String,
    val menuBarStyle: String,
    val primaryMetric: String,
    val secondaryMetric: String,
    val detailStyle: String,
    val textSize: String,
)

@Serializable
data class DeviceSyncNotifications(
    val claudeSession: Int,
    val claudeSevenDay: Int,
    val claudeFable: Int,
    val openAIWeekly: Int,
    val openAIResetCredits: Int,
    val cursorAPI: Int,
    val cursorAuto: Int,
    val cursorCredit: Int,
)

@Serializable
data class DeviceSyncConnections(
    val openAISessionToken: String? = null,
    val cursorSessionToken: String? = null,
    val elevenLabsAPIKey: String? = null,
) {
    val count: Int
        get() = listOf(openAISessionToken, cursorSessionToken, elevenLabsAPIKey)
            .count { !it.isNullOrBlank() }
}

@Serializable
data class DeviceSyncPayload(
    val version: Int,
    val issuedAtEpochSeconds: Long,
    val expiresAtEpochSeconds: Long,
    val general: DeviceSyncGeneral? = null,
    val appearance: DeviceSyncAppearance? = null,
    val notifications: DeviceSyncNotifications? = null,
    val connections: DeviceSyncConnections? = null,
)

data class DevicePairingCode(
    val sessionID: String,
    val host: String,
    val port: Int,
    val desktopID: String,
    val desktopName: String,
    val desktopPublicKey: ByteArray,
)

@Serializable
data class DevicePairRequest(
    val sessionID: String,
    val deviceID: String,
    val deviceName: String,
    val publicKey: String,
)

@Serializable
data class DevicePairStartResponse(
    val status: String,
    val confirmationCode: String,
)

@Serializable
data class DevicePairPollResponse(
    val status: String,
    val desktopID: String? = null,
    val desktopName: String? = null,
    val envelope: DeviceEncryptedEnvelope? = null,
    val message: String? = null,
)

@Serializable
data class DeviceEncryptedEnvelope(
    val nonce: String,
    val ciphertext: String,
    val tag: String,
)

@Serializable
data class DeviceStatusCommand(
    val action: String,
    val issuedAtEpochSeconds: Long,
)

@Serializable
data class DeviceWipeAcknowledgement(
    val desktopID: String,
    val deviceID: String,
    val timestamp: Long,
    val proof: String,
)

object DeviceSyncCodec {
    const val PAIRING_INFO = "agentusagebar-device-pair-v2"
    const val STATUS_INFO = "agentusagebar-device-status-v2"
    const val CURRENT_PAIRING_VERSION = 2
    const val CURRENT_PAYLOAD_VERSION = 1

    val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    fun decodePairingCode(rawValue: String): DevicePairingCode {
        require(rawValue.length <= 4_096) { "This QR code is too large." }
        val uri = runCatching { URI(rawValue) }
            .getOrElse { throw IllegalArgumentException("This is not an Agent Usage Bar pairing code.") }
        require(uri.scheme == "agentusagebar" && uri.host == "pair" && uri.path == "/v2") {
            "This is not an Agent Usage Bar pairing code."
        }
        val query = queryItems(uri.rawQuery)
        require(query["v"]?.toIntOrNull() == CURRENT_PAIRING_VERSION) {
            "Update Agent Usage Bar to use this pairing code."
        }
        val session = query["session"] ?: error("Pairing session is missing.")
        val host = query["host"] ?: error("Mac network address is missing.")
        val port = query["port"]?.toIntOrNull()?.takeIf { it in 1..65_535 }
            ?: error("Pairing port is invalid.")
        val desktop = query["desktop"] ?: error("Mac identity is missing.")
        val name = query["name"] ?: "Mac"
        val key = query["key"]?.let(::base64URLDecode)
            ?: error("Mac public key is missing.")
        require(key.size == 65 && key.first() == 4.toByte()) {
            "Mac public key is invalid."
        }
        return DevicePairingCode(session, host, port, desktop, name, key)
    }

    fun decodePayload(data: ByteArray, nowSeconds: Long = System.currentTimeMillis() / 1_000): DeviceSyncPayload {
        val payload = runCatching {
            json.decodeFromString<DeviceSyncPayload>(data.toString(Charsets.UTF_8))
        }.getOrElse { throw IllegalArgumentException("The transferred settings are damaged.") }
        require(payload.version == CURRENT_PAYLOAD_VERSION) {
            "Update Agent Usage Bar to import these settings."
        }
        require(payload.expiresAtEpochSeconds >= nowSeconds) {
            "The pairing transfer expired. Generate a new code on your Mac."
        }
        return payload
    }

    fun base64URLEncode(value: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(value)

    fun base64URLDecode(value: String): ByteArray = Base64.getUrlDecoder().decode(value)

    fun credentialHash(value: String?): String? = value
        ?.takeIf { it.isNotBlank() }
        ?.let { base64URLEncode(MessageDigest.getInstance("SHA-256").digest(it.toByteArray())) }

    private fun queryItems(rawQuery: String?): Map<String, String> =
        rawQuery.orEmpty().split("&").mapNotNull { item ->
            val split = item.indexOf('=')
            if (split < 0) null else {
                URLDecoder.decode(item.substring(0, split), StandardCharsets.UTF_8.name()) to
                    URLDecoder.decode(item.substring(split + 1), StandardCharsets.UTF_8.name())
            }
        }.toMap()
}

object DeviceSyncCrypto {
    fun generateDeviceKeyPair(): KeyPair =
        KeyPairGenerator.getInstance("EC").apply {
            initialize(ECGenParameterSpec("secp256r1"))
        }.generateKeyPair()

    fun restorePrivateKey(encoded: String): PrivateKey =
        KeyFactory.getInstance("EC").generatePrivate(
            PKCS8EncodedKeySpec(DeviceSyncCodec.base64URLDecode(encoded)),
        )

    fun rawPublicKey(publicKey: PublicKey): ByteArray {
        val ec = publicKey as ECPublicKey
        return byteArrayOf(4) +
            fixedWidth(ec.w.affineX, 32) +
            fixedWidth(ec.w.affineY, 32)
    }

    fun parseRawPublicKey(raw: ByteArray): PublicKey {
        require(raw.size == 65 && raw[0] == 4.toByte()) { "Invalid P-256 public key." }
        val parameters = AlgorithmParameters.getInstance("EC").apply {
            init(ECGenParameterSpec("secp256r1"))
        }.getParameterSpec(ECParameterSpec::class.java)
        val point = ECPoint(
            BigInteger(1, raw.copyOfRange(1, 33)),
            BigInteger(1, raw.copyOfRange(33, 65)),
        )
        return KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(point, parameters))
    }

    fun sharedSecret(privateKey: PrivateKey, peerPublicKey: ByteArray): ByteArray =
        KeyAgreement.getInstance("ECDH").run {
            init(privateKey)
            doPhase(parseRawPublicKey(peerPublicKey), true)
            generateSecret()
        }

    fun deriveKey(sharedSecret: ByteArray, salt: String, info: String): ByteArray {
        val extract = Mac.getInstance("HmacSHA256").run {
            init(SecretKeySpec(salt.toByteArray(), "HmacSHA256"))
            doFinal(sharedSecret)
        }
        return Mac.getInstance("HmacSHA256").run {
            init(SecretKeySpec(extract, "HmacSHA256"))
            doFinal(info.toByteArray() + byteArrayOf(1)).copyOf(32)
        }
    }

    fun confirmationCode(sharedSecret: ByteArray, sessionID: String): String {
        val key = deriveKey(sharedSecret, sessionID, DeviceSyncCodec.PAIRING_INFO)
        val number = (
            ((key[0].toLong() and 0xff) shl 24) or
                ((key[1].toLong() and 0xff) shl 16) or
                ((key[2].toLong() and 0xff) shl 8) or
                (key[3].toLong() and 0xff)
            ) % 1_000_000
        return number.toString().padStart(6, '0')
    }

    fun authenticationProof(
        sharedSecret: ByteArray,
        salt: String,
        info: String,
        message: String,
    ): String {
        val key = deriveKey(sharedSecret, salt, info)
        val proof = Mac.getInstance("HmacSHA256").run {
            init(SecretKeySpec(key, "HmacSHA256"))
            doFinal(message.toByteArray())
        }
        return DeviceSyncCodec.base64URLEncode(proof)
    }

    fun open(
        envelope: DeviceEncryptedEnvelope,
        sharedSecret: ByteArray,
        salt: String,
        info: String,
    ): ByteArray {
        val key = deriveKey(sharedSecret, salt, info)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(key, "AES"),
            GCMParameterSpec(128, DeviceSyncCodec.base64URLDecode(envelope.nonce)),
        )
        return cipher.doFinal(
            DeviceSyncCodec.base64URLDecode(envelope.ciphertext) +
                DeviceSyncCodec.base64URLDecode(envelope.tag),
        )
    }

    private fun fixedWidth(value: BigInteger, size: Int): ByteArray {
        val bytes = value.toByteArray()
        return when {
            bytes.size == size -> bytes
            bytes.size > size -> bytes.copyOfRange(bytes.size - size, bytes.size)
            else -> ByteArray(size - bytes.size) + bytes
        }
    }
}
